unicodetools/src/main/java/org/unicode/utilities/LinkUtilities.java

```diff
@@ -0,0 +1,1146 @@
+        }
+
+        static Part fromInitiator(int cp) {
+            for (Part part : Part.values()) {
+                if (part.initiator == cp) {
+                    return part;
+                }
+            }
+            return null;
+        }
+
+        /**
+         * Pull apart a URL string into Parts. <br>
+         * TODO: unescape the %escapes.
+         *
+         * @param source
+         * @param unescape TODO
+         * @return
+         */
+        public static NavigableMap<Part, String> getParts(String source, boolean unescape) {
+            Map<Part, String> result = new HashMap<>();
+            // quick and dirty
+            int partStart = 0;
+            int partEnd;
+            main:
+            for (Part part : Part.values()) {
+                switch (part) {
+                    case PROTOCOL:
+                        partEnd = source.indexOf("://"); // TODO fix for mailto
+                        if (partEnd > 0) {
+                            partEnd += 3;
+                            result.put(Part.PROTOCOL, source.substring(0, partEnd));
+                            partStart = partEnd;
+                        }
+                        break;
+                    default:
+                        partEnd =
+                                part.terminators.span(
+                                        source, partStart, SpanCondition.NOT_CONTAINED);
+                        if (partStart != partEnd) {
+                            result.put(part, part.unescape(source.substring(partStart, partEnd)));
+                        }
+                        if (partEnd == source.length()) {
+                            break main;
+                        }
+                        partStart = partEnd;
+                        break;
+                }
+            }
+            return ImmutableSortedMap.copyOf(result);
+        }
+
+        /**
+         * Unescape a part. But don't unescape interior characters or terminators because they are
+         * content! For example "a/b%2Fc" as a path should not be turned into a/b/c, because that
+         * b/c is a path-part.
+         *
+         * @param substring
+         * @return
+         */
+        public String unescape(String substring) {
+            return LinkUtilities.unescape(substring, extraQuoted);
+        }
+
+        static final int[] FRAGMENT_DIRECTIVE = ":~:".codePoints().toArray();
+    }
+
+    private static final UnicodeSet idnMapped =
+            IUP.getSet("Idn_Status=" + Idn_Status_Values.mapped);
+    private static final UnicodeSet idnValid = IUP.getSet("Idn_Status=" + Idn_Status_Values.valid);
+    static final Pattern protocol = Pattern.compile("(https?://|mailto:)");
+    public static final UnicodeSet validHost =
+            new UnicodeSet(idnValid)
+                    .addAll(idnMapped)
+                    .removeAll(new UnicodeSet("[:Block=Basic_Latin:]"))
+                    .addAll(new UnicodeSet("[-a-zA-Z0-9..]"))
```

Comment:
why two dots?

Reply:
Fixed this. Tho not currently used by the tests (because we don't focus on the domain name) the dots ".．。" are treated as identical in toUTS46n.

Reply:
Didn't resolve this yet to give you a chance to see the change.